plugins/commands/cloud/provider/create.rb

```diff
@@ -0,0 +1,72 @@
+require 'optparse'
+
+module VagrantPlugins
+  module CloudCommand
+    module ProviderCommand
+      module Command
+        class Create < Vagrant.plugin("2", :command)
+          def execute
+            options = {}
+
+            opts = OptionParser.new do |o|
+              o.banner = "Usage: vagrant cloud provider create [options] organization/box-name provider-name version [url]"
+              o.separator ""
+              o.separator "Creates a provider entry on Vagrant Cloud"
+              o.separator ""
+              o.separator "Options:"
+              o.separator ""
+
+              o.on("-u", "--username USERNAME_OR_EMAIL", String, "Specify your Vagrant Cloud username or email address") do |t|
+                options[:username] = u
+              end
+            end
+
+            # Parse the options
+            argv = parse_options(opts)
+            return if !argv
+            if argv.empty? || argv.length > 4
+              raise Vagrant::Errors::CLIInvalidUsage,
+                help: opts.help.chomp
+            end
+
+            @client = VagrantPlugins::CloudCommand::Util.client_login(@env, options[:username])
+
+            box = argv.first.split('/')
+            org = box[0]
+            box_name = box[1]
+            provider_name = argv[1]
+            version = argv[2]
+            url = argv[3]
+
+            upload_provider(org, box_name, provider_name, version, url, @client.token, options)
+          end
+
+          def upload_provider(org, box_name, provider_name, version, url, access_token, options)
+            if !url
+              @env.ui.warn(I18n.t("cloud_command.upload.no_url"))
+            end
+
+            org = options[:username] if options[:username]
+
+            server_url = VagrantPlugins::CloudCommand::Util.api_server_url
+            account = VagrantPlugins::CloudCommand::Util.account?(org, access_token, server_url)
+            box = VagrantCloud::Box.new(account, box_name, nil, nil, nil, access_token)
+            cloud_version = VagrantCloud::Version.new(box, version, nil, nil, access_token)
+            provider = VagrantCloud::Provider.new(cloud_version, provider_name, nil, url, org, box_name, access_token)
```

Comment:
A note on this....basically every subcommand follows this pattern of creating all these `VagrantCloud` objects to make API requests with the gem. It gets bad for a class like `Provider` that needs every single object. I thought about maybe abstracting it out into the helper class, but that would involve a large switch case depending on which API call the requester wants (i.e. a Version subcommand would only create Box and Version class objects and return the Version object, but Box would only create and return Box). Maybe that's the right move though?